### PR TITLE
Call all monitors using 'number of completed steps' as `iter`

### PR DESCRIPTION
**Pull request by johnomotani**
Previously, the `Solver` implementations passed the loop counter to the `iter` argument of `call_monitors`, but since the iteration has already been completed when the monitors are called, this results in `iter` always being one less than the number of completed monitor-steps at the point when the moniters are called, which is confusing.

Technically non-backward-compatible, but I'd argue the existing behaviour is a bug and guess it's unlikely anyone is relying on it (if anyone was, they'd probably have complained already!)?

The `PetscSolver` implementation is unmodified, because I think from a quick glance at the code (though haven't actually checked) that it follows the 'updated' behaviour already.

Note: this PR is a merge into #2534, as it is a proposal to address a question of how far that 'fix' should go.

### CODE REVIEW

File: src/solver/impls/arkode/arkode.cxx [hunk 1]
```diff
@@ -545,7 +545,7 @@ int ArkodeSolver::run() {
 
     /// Call the monitor function
 
-    if (call_monitors(simtime, i, NOUT)) {
+    if (call_monitors(simtime, i+1, NOUT)) {
```
[the remark this answers is not in the document]
Comment (johnomotani) [reply src/solver/impls/arkode/arkode.cxx:548]: @ZedThree do we want to fix these while we're touching this code (if we are...)?

**Comment by johnomotani** [comment]
I don't feel strongly about this change. This PR wouldn't make any difference to me (#2535 is the one I care about, and even there only really the `hist_hi` changes, not the `iteration changes). I thought it might make things more consistent, but if it does actually break code, I'm happy to forget about it!

**Comment by ZedThree** [comment]
I've been bitten by this before, so I'm generally in favour of it. The change in `solver.cxx` illustrates the non-obviousness of the current approach quite clearly: we need to pass `-1` for the initial iteration.

But if there is code that relies on it, then it probably shouldn't go into master

**Comment by johnomotani** [comment]
Sounds like the best plan might be to re-jig this to go into `next`, once #2534 has gone into `master` and then merged into `next`.

**Comment by bendudson** [comment]
Is this now ready to go into `next`, since #2534 has been merged? Or is it still a work in progress?

**Comment by johnomotani** [comment]
I think it still needs to wait for `master` to be merged into `next` to avoid conflicts.

I think the logic of what value excatly the 'iteration count' should have needs to be checked and discussed, e.g in terms of the value of `solver->iteration` compared to the `iter` argument to `Monitor::call()`. At minimum, the description probably needs updating here https://github.com/boutproject/BOUT-dev/blob/ab69d52f362dcaaddb22f3bd3b16aa7bab924d84/include/bout/solver.hxx#L432-L448

**Comment by dschwoerer** [comment]
Replaced by #2882

**Pull request closed without merging**